### PR TITLE
refactor `dynamic/fast` - use 'slices'

### PR DESCRIPTION
**Pull request by OllysCoding**
This PR refactors `dynamic/fast` to use the conventions as other dynamic type containers (e.g [dynamic/slow](https://github.com/guardian/dotcom-rendering/pull/5892)). 

`dynamic/fast` was originally built in https://github.com/guardian/dotcom-rendering/pull/5427 - A first attempt at supporting dynamic containers, groups & boosting, it mostly worked but the code is hard to work on & understand.

---

This PR includes the refactor of dynamic/fast to use the 'slice' convention, and fixes some accompanying bugs in dynamic/fast & dynamic/slow - ensuring that font sizes are correct, sublinks show on the correct cards, etc.

### Using slices

As seen in other PRs, this refactor introduces the separation of the grouping of cards & layout decisions, and the actual rendering itself. I'd describe this as a 'soft' separation, as [unlike frontend](https://github.com/guardian/frontend/blob/main/common/app/layout/slices/Slice.scala), the layout logic & rendering stay close together, making it easy to work on.

The basic pattern looks like (not accurate to how dynamic/fast works):

```ts
const Card25_Card25_Card25_Card25 = ({ cards }) => {
    // This is responsible for rendering a row (slice) of 4 cards, each with take up 25% of the container
}

const Card25_Card25_ColumnOfThreeCards25_ColumnOfThreeCards25 = ({ cards }) => {
    // This is responsible for rendering a slice of cards, where the first 2 take up 25% width, 
    // and alongside that there are 2 columns of cards each which take 3 cards.
}

// ... Rest of the slice components are defined here

const DynamicFast = ({ groups }) => {
    // Define a layout type & an array of cards, we'll then use logic to decide which layout is correct,
    // and create our cards array to use the correct cards for that layout.
    let layout: "layout1" | "layout2" | "layout3" | "layout4" = undefined;
    let cards = []

    // Looking at the cards we've received from fronts tool / facia press, 
    // now grouped into 'huge', 'veryBig', 'big' & 'standard'
    // we can make an informed decision on which layout we want, and orchestrate the cards accordingly
    if (groups.huge.length > 3) {
        layout = "layout1"
        cards = [...groups.huge.slice(0, 4)]
    } else if (groups.big.length > 1 && groups.standards.length > 6) {
         layout = "layout2"
         cards = [...groups.big.slice(0, 2), ...groups.standard.slice(0, 6)]
    }  // ...
    
    // Now we simply take the layout & cards array, and pass it to the accompanying component!
    switch(layout) {
        case "layout1"
            return <Card25_Card25_Card25_Card25 cards={cards} />
        case "layout2"
            return <Card25_Card25_ColumnOfCards25_ColumnOfCards25 cards={cards} />
        // ...
    }
}

```

--- 

This PR also tidies up the cards themselves, introducing a number of fixes:
- introduces `oneVeryBig` layout - previously this showed as oneHuge: https://github.com/guardian/dotcom-rendering/issues/5782
- Share first slice components between dynamic/fast & dynamic/slow
- switch from explicitly defining `secondSliceGroupedTrails` to filtering them out after the first slice is defined
- Make sure sublinks show on the correct cards on dynamic/slow
- Make sure the font sizes are correct for all slices
- Refactor stories on dynamic/slow to include the slice layouts in the container description
- Create new stories on dynamic/fast which cover all layouts & edge cases

### Showing slices in storybook

To improve developer experience, I update all dynamic/fast and dynamic/slow stories to include the slice layout names in the stories:

image

### First Slice

| Name      | Screenshot      |
|-------------|------------|
| 1 huge | ![1huge][] |
| 1 very big | ![1verybig][] |
| 2 very bigs | ![2verybigs][] |
| 2 very bigs (1st boosted) | ![1stboosted][] |
| 2 very bigs (2nd boosted) | ![2ndboosted][] |

[1huge]: https://user-images.githubusercontent.com/9575458/194859535-65accecf-551c-4bc2-b77d-3972da02ba68.png
[1verybig]: https://user-images.githubusercontent.com/9575458/194859595-9de44735-de7e-49aa-a8f4-ec1b75406611.png
[2verybigs]: https://user-images.githubusercontent.com/9575458/194859651-62492255-9ef5-4d61-aca6-ad0664c5838c.png
[1stboosted]: https://user-images.githubusercontent.com/9575458/194859710-fdfc2fd9-265c-4de0-88fd-c940c50c92e9.png
[2ndboosted]: https://user-images.githubusercontent.com/9575458/194859780-26b71461-c7a7-4aed-8d84-ea22f1e26250.png

### Second Slice

| Name      | Screenshot      |
|-------------|------------|
| 2 or more bigs (1st boosted) | ![twoOrMoreBigsFirstBoosted][] |
| 4 bigs | ![fourBigs][] |
| 3 bigs | ![threeBigs][] |
| 2 bigs | ![twoBigs][] |
| 1 big | ![oneBig][] |
| 1 big (boosted) | ![oneBigBoosted][] |
| 0 bigs | ![noBigs][] |

[twoOrMoreBigsFirstBoosted]: https://user-images.githubusercontent.com/9575458/194860690-af35965e-a0d8-4e69-8c14-27c4fa482041.png

[fourBigs]: https://user-images.githubusercontent.com/9575458/194860421-e98c90f3-5642-4bb5-807d-18b415ffae85.png

[threeBigs]: https://user-images.githubusercontent.com/9575458/194860374-a87a434b-c1bf-4003-91b1-02f4ad0ce9f2.png

[twoBigs]: https://user-images.githubusercontent.com/9575458/194860821-a9474d5e-9d18-442b-85c3-dcd01e3a6a7a.png

[oneBig]: https://user-images.githubusercontent.com/9575458/194860078-4673a970-26de-4d8e-b4b2-2efd072a0fdc.png
[oneBigBoosted]: https://user-images.githubusercontent.com/9575458/194860931-a95281f3-be1c-48c1-843e-847cd746b04a.png
[noBigs]: https://user-images.githubusercontent.com/9575458/194859865-01db5f16-1c05-410f-b3d7-cccf07f3dbce.png

### CODE REVIEW

File: dotcom-rendering/src/web/components/DynamicFast.tsx
Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/DynamicFast.tsx:526]: Failing this, maybe there could be `secondSliceBigs` and `secondSliceStandards`? I think I get tripped up on there being another intermediary grouped object

File: dotcom-rendering/src/web/components/DynamicFast.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/DynamicFast.tsx:758]: This and the logic above it leading into it is really great

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/lib/dynamicSlices.tsx:77]: This can be something for another PR if it is a thing but do these need to be set in such a way that they don't both show at the same time?

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/lib/dynamicSlices.tsx:139]:
```suggestion
/* ._________________.
 * |#################|
 * |#################|
 * |#################|
 * |_________________|
 */
```

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx
Comment (oliverlloyd) [inline dotcom-rendering/src/web/lib/dynamicSlices.tsx:169]:
```suggestion
/* .___________________________________.
```

**Review by oliverlloyd — APPROVED** [review]
Impressive work!

I've made some comments but nothing huge, big or very big.

It's a lot of changes so I can't say, hand on heart, that I've checked everything here but the general approach is fantastic and the code diffs that I have checked look sound so I'm happy to approve this, leave any updates based on comments up to you and move forward

File: dotcom-rendering/src/web/components/DynamicFast.tsx
Comment (bryophyta) [inline dotcom-rendering/src/web/components/DynamicFast.tsx:293]: a small point, but: do we need a `key` here, given that we're not inside a loop?

File: dotcom-rendering/src/web/components/DynamicFast.tsx
Comment (jamesgorrie) [reply dotcom-rendering/src/web/components/DynamicFast.tsx:203]: Really useful for when you use `.entries` e.g.
```js
for (const [value, index] of arr.entries()) {
  // blah blah.
}
```

**Review by bryophyta — APPROVED** [review]
This looks great to me! I've put a couple of small comments, but nothing blocking.

The one more significant thing I was wondering: With changes to such a complicated component, it feels like visual regression tests are especially important given how hard it is to work out the rendering logic in your head as a reviewer. It looks like the storybook stories have been changed quite significantly at the same time as some of the rendering code in this PR. The storybook changes look like great additions! But with my limited knowledge of Chromatic -- does this leave us without much visual regression coverage for this particular PR?

File: dotcom-rendering/src/web/components/DynamicFast.tsx
Comment (mxdvl) [reply dotcom-rendering/src/web/components/DynamicFast.tsx:203]: I also had this realisation in 35e4393db0c9141ab0e87f0153f7830a9a676ff3. However, having another think about this, I’m not entirely sure it’s an improvement over using `remaining.length`.

It’s definitely neater this way, but there could be issues with maintainability, as it requires developers to be confident with both destructuring and the order of map arguments.

Comment (oliverlloyd) [reply dotcom-rendering/src/web/components/DynamicFast.tsx:203]: +1 to Max's point. I think most developers know the first two properties from `map` (item, index) but the fact the third property is a copy of the original array is less known and could be confusing? Also, `remaining.length` just reads so _nice_, shame to waste that

File: dotcom-rendering/src/web/components/DynamicFast.tsx
Comment (bryophyta) [inline dotcom-rendering/src/web/components/DynamicFast.tsx:639]: if `secondSliceLayout` doesn't get set in this switch block then it can create a runtime error when `SecondSlice` switches on `secondSliceLayout` below, because the latter switch doesn't have a default case.

`tsc` isn't helping us here (not sure if this is because the assignment is complex, or because `secondSliceLayout` is initialised as `undefined`?

I think the code as written works, but with such a complex set of conditions I feel a bit uneasy about not being able to rely on the compiler/tooling to tell us whether we're actually catching all of the different cases. To prevent the bug, we could just add a default case on line 752 in this file, but it would be nice if  we could restructure to allow the compiler to tell us when we risk not setting a value for `secondSliceLayout`. I realise that mutation is well-motivated here, but wondering if there's a way around it? (Or a different way of narrowing the type before it gets used?)

File: dotcom-rendering/src/web/components/DynamicFast.tsx
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/DynamicFast.tsx:84]: I'm super glad you asked - I think the problems before came from padding, margins, or stretching. However it seems like this works now (likely thanks to @mxdvl's changes to use the `gap` property).

Now refactored to use just LI's & UL's :D

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx
Comment (OllysCoding) [reply dotcom-rendering/src/web/lib/dynamicSlices.tsx:77]: Testing this;

with trail text & supporting content set on both, the 75 card keeps both, the 25 card only has supporting content:
image

I'll look into this!

File: dotcom-rendering/src/web/components/DynamicFast.tsx
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/DynamicFast.tsx:165]: I think I'd definitely like to see this refactored in a separate PR, there's definitely room for improvement here - the way it works is quite convoluted!

**Comment by OllysCoding** [comment]
> 🤔 Is the shouldPadWrappableRows function perhaps similar to [this one](https://github.com/guardian/dotcom-rendering/blob/622bdea81332c99a5fd07cf9702a256095b44067/dotcom-rendering/src/web/components/FixedMediumFastXI.tsx#L18)?

It looks like this is the case! I think we might want to look at a proper refactor for this later - I don't want to make a change to this area as part of this PR (this PR currently just replicates the existing functionality), but I think we need to have someone take a step back and look at all of the dividers, etc, and come up with a more unified approach

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx
Comment (OllysCoding) [reply dotcom-rendering/src/web/lib/dynamicSlices.tsx:77]: Fixed :)

**Comment by oliverlloyd** [comment]
> I think we need to have someone take a step back and look at all of the dividers, etc, and come up with a more unified approach

👍  looking at this later post both PRs make sense

File: dotcom-rendering/src/web/components/DynamicFast.tsx
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/DynamicFast.tsx:164]: Fixed :)

File: dotcom-rendering/src/web/components/DynamicFast.tsx
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/DynamicFast.tsx:584]: +1 I've added a default case to `SecondSlice` - it's frustrating tsc doesn't pick up on this :(

**Comment by bryophyta** [comment]
> 🤔 Is the `shouldPadWrappableRows` function perhaps similar to [this one](https://github.com/guardian/dotcom-rendering/blob/622bdea81332c99a5fd07cf9702a256095b44067/dotcom-rendering/src/web/components/FixedMediumFastXI.tsx#L18)?

fwiw I've got an idea for simplifying at least some of the divider functionality that I discussed with @OllysCoding the other day. I was waiting until this PR was merged to start work on it though. If someone else is planning on taking a look at this then I'm happy to jot down my thoughts and/or pair on this. Alternatively I'm happy to raise a PR for my proposed refactor and see what people think of it.

File: dotcom-rendering/src/web/components/DynamicFast.tsx
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/DynamicFast.tsx]: I've gone with an improved solution where we create the `secondSliceGroupedTrails` by filtering out the cards used in the first slice - I think this reads a lot better!

Comment (OllysCoding) [reply dotcom-rendering/src/web/components/DynamicFast.tsx]:
```ts
// Create our object of grouped trails that doesn't include any
// cards used by the first slice
const secondSliceGroupedTrails = filterGroupedTrails({
	groupedTrails,
	filter: firstSliceCards,
});
```

**Pull request merged**
